src/sources/opentelemetry/http.rs

```diff
@@ -0,0 +1,155 @@
+use std::net::SocketAddr;
+
+use bytes::Bytes;
+use futures_util::FutureExt;
+use http::StatusCode;
+use prost::Message;
+use snafu::Snafu;
+use tracing::Span;
+use vector_core::{
+    event::{BatchNotifier, BatchStatus},
+    ByteSizeOf,
+};
+use warp::{filters::BoxedFilter, reject::Rejection, reply::Response, Filter, Reply};
+
+use crate::{
+    event::Event,
+    internal_events::{EventsReceived, HttpBytesReceived, StreamClosedError},
+    opentelemetry::LogService::{ExportLogsServiceRequest, ExportLogsServiceResponse},
+    shutdown::ShutdownSignal,
+    sources::util::{decode, ErrorMessage},
+    tls::MaybeTlsSettings,
+    SourceSender,
+};
+
+use super::{reply::protobuf, status::Status};
+
+#[derive(Clone, Copy, Debug, Snafu)]
+pub(crate) enum ApiError {
+    BadRequest,
+    ServerShutdown,
+}
+
+impl warp::reject::Reject for ApiError {}
+
+pub(crate) async fn run_http_server(
+    address: SocketAddr,
+    tls_settings: MaybeTlsSettings,
+    filters: BoxedFilter<(Response,)>,
+    shutdown: ShutdownSignal,
+) -> crate::Result<()> {
+    let span = Span::current();
+    let listener = tls_settings.bind(&address).await?;
+    let routes = filters
+        .with(warp::trace(move |_info| span.clone()))
+        .recover(|r: Rejection| async move {
+            if let Some(e_msg) = r.find::<ErrorMessage>() {
+                let reply = protobuf(Status {
+                    code: 2, // UNKNOWN - OTLP doesn't require use of status.code, but we can't encode a None here
+                    message: e_msg.message().into(),
+                    ..Default::default()
+                });
+
+                Ok(warp::reply::with_status(reply, e_msg.status_code()))
+            } else {
+                // other internal error - will return 500 internal server error
+                error!(message = "Unhandled rejection.", rejection = ?r);
+                Err(r)
+            }
+        });
+
+    info!(message = "Building HTTP server.", address = %address);
+
+    warp::serve(routes)
+        .serve_incoming_with_graceful_shutdown(listener.accept_stream(), shutdown.map(|_| ()))
+        .await;
+
+    Ok(())
+}
+
+pub(crate) fn build_warp_filter(
+    acknowledgements: bool,
+    out: SourceSender,
+    protocol: &'static str,
+) -> BoxedFilter<(Response,)> {
+    warp::post()
+        .and(warp::path!("v1" / "logs"))
+        .and(warp::header::exact_ignore_case(
+            "content-type",
+            "application/x-protobuf",
+        ))
+        .and(warp::header::optional::<String>("content-encoding"))
+        .and(warp::body::bytes())
+        .and_then(move |encoding_header: Option<String>, body: Bytes| {
+            emit!(HttpBytesReceived {
+                byte_size: body.len(),
+                http_path: "/v1/logs",
+                protocol,
+            });
```

Comment:
I believe this is supposed to be _after_ decompression? Also, are we missing `HttpEventsReceived`? Or is that emitted by some layer?

Reply:
`HttpEventsReceived` 🤔 the test assertion seems to think we are... and I don't see it in the `datadog_agent` code either: https://cs.github.com/vectordotdev/vector?q=path%3A%2F%5Esrc%5C%2Fsources%5C%2Fdatadog_agent%2F+HttpEventsReceived

Reply:
Both `opentelemetry` and `datadog_agent` do emit `EventsReceived` - https://cs.github.com/vectordotdev/vector?q=path%3A%2F%5Esrc%5C%2Fsources%5C%2Fdatadog_agent%2F+EventsReceived

Reply:
We may want to double check the `datadog_agent` is emitting the right events?

Fixed the `EventsReceived` I had here to be `HttpEventsReceived` in 566849ba7e7876cbfdefc56078d1e2422ef6857f